packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart

```diff
@@ -0,0 +1,150 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:math';
+
+import 'package:process/process.dart';
+
+import '../../artifacts.dart';
+import '../../base/error_handling_io.dart';
+import '../../base/file_system.dart';
+import '../../base/io.dart';
+import '../../base/logger.dart';
+import '../../flutter_manifest.dart';
+import '../build_system.dart';
+
+/// Applies a series of user-specified asset-transforming packages to an asset file.
+final class AssetTransformer {
+  AssetTransformer({
+    required ProcessManager processManager,
+    required Logger logger,
+    required FileSystem fileSystem,
+    required Artifacts artifacts,
+  })  : _processManager = processManager,
+        _logger = logger,
+        _fileSystem = fileSystem {
+          _dartBinary = artifacts.getArtifactPath(Artifact.engineDartBinary);
+        }
+
+  final ProcessManager _processManager;
+  final Logger _logger;
+  final FileSystem _fileSystem;
+  late String _dartBinary;
```

Comment:
`late` fields should be avoided (it's a hole in static type safety). late final with initializer expressions are safer, though you'd need to make the artifacts passed to the constructor a private field on the class so you can access it here.

Reply:
So:

```dart
  final Artifacts _artifacts;
  late final String _dartBinary = _artifacts.getArtifactPath(Artifact.engineDartBinary);
```

Reply:
Addressed in [refactor AssetTransformer class to avoid the need for a late-initia…](https://github.com/flutter/flutter/pull/143815/commits/16a21086d33ca96e21d830e64584d4d6a8d85655)